protocols/mdns/src/behaviour/iface.rs

```diff
@@ -198,43 +199,51 @@ impl InterfaceState {
         params: &impl PollParameters,
     ) -> Option<(PeerId, Multiaddr, Instant)> {
         // Poll receive socket.
-        while self.recv_socket.poll_readable(cx).is_ready() {
-            match self
-                .recv_socket
-                .recv_from(&mut self.recv_buffer)
-                .now_or_never()
-            {
-                Some(Ok((len, from))) => {
+        match self.recv_socket.poll_read(cx, &mut self.recv_buffer) {
+            Poll::Ready(result) => match result {
+                Ok((len, from)) => {
                     if let Some(packet) = MdnsPacket::new_from_bytes(&self.recv_buffer[..len], from)
                     {
                         self.inject_mdns_packet(packet, params);
                     }
                 }
-                Some(Err(err)) => log::error!("Failed reading datagram: {}", err),
-                None => {}
-            }
+                Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                    // Not more bytes available on the socket to read
+                }
+                Err(err) => {
+                    log::error!("failed reading datagram: {}", err);
+                }
+            },
+            Poll::Pending => {}
         }
+
         // Send responses.
-        while self.send_socket.poll_writable(cx).is_ready() {
-            if let Some(packet) = self.send_buffer.pop_front() {
-                match self
-                    .send_socket
-                    .send_to(&packet, SocketAddr::new(self.multicast_addr, 5353))
-                    .now_or_never()
-                {
-                    Some(Ok(_)) => log::trace!("sent packet on iface {}", self.addr),
-                    Some(Err(err)) => {
-                        log::error!("error sending packet on iface {}: {}", self.addr, err)
+        while let Some(packet) = self.send_buffer.pop_front() {
+            match self.send_socket.poll_write(
+                cx,
+                &packet,
+                SocketAddr::new(self.multicast_addr, 5353),
+            ) {
+                Poll::Ready(data) => match data {
+                    Ok(_) => log::trace!("sent packet on iface {}", self.addr),
+                    Err(err) => {
+                        log::error!("error sending packet on iface {} {}", self.addr, err);
+                        self.send_buffer.push_front(packet);
```

Comment:
This is not functionally equivalent to what we did before.

Previously, we discarded a packet that we failed to send AND moved on to the next one if the socket was still ready to accept packets.

Now we are pushing the packet back into the buffer and breaking from the loop.